HAL.md

```diff
@@ -0,0 +1,115 @@
+# Hardware abstract layer(HAL) overview
+
+## Abstract
+Hardware abstract layer(HAL) defines a standard interface for hardware vendors to implement, which enables Tangle-Accelerator to be aware of lower-level driver implementations. HAL layer supports such as uart, GPIO, secure storage and other low-level operators inside endpoint. The endpoint is a daemon based on Linux which offers monitor for the device and can send the message to tangle-accelerator. HAL layer lets the user only need to implement their device operators from abstract interface which defined inside `device.h` but not need to know how endpoint work.
+
+## Directory overview
+* connectivity: Connect to Tangle-accelerator functions.
+* devices: The real directory to implement devices.
+* hal: Hardware abstraction layer.
+* hooks: Git hook functions.
+* tests: Unit-tests.
+* third_party: Third party directory.
+* utils: Utility functions for TA-endpoint.
+
+## Implement new device
+
+### Create directory for new device
+Create directory for new device into `devices`.
+```
+mkdir devices/mydevice
+```
+* Create `impl.c` and `Makefile` under new device directory.
+* Include `device.h` headers into new device header file.
+
+Implement new device operations, which are included from `hal/device.h`
+
+### Abstraction
+Here are some operations needed to implement inside new device:
+* device_operations
+    * init : initialize device
+    * fini : finalize device
+    * get_key : get device key
+    * get_device_id : get device id(IMEI or other identifier)
+* uart_operations
+    * init : initialize uart
+    * write : write command to uart device
+    * read : read from uart device
+    * clean : flush buffer
+* secure_store_operations
+    * init : initialize secure storage
+    * write : write item to secure storage
+    * read : read item from secure storage
+    * delete : delete item inside secure storage
+### Functions to use inside `impl.c`
```

Comment:
Delete the sections "Functions to use inside impl.c", "Modify hal Makefile", "Implement operators inside impl.c", etc. We should write down high-level descriptions first.

Reply:
Let's concentrate on qualified interface for further implementation.